beacon-chain/core/altair/epoch_precompute.go

```diff
@@ -0,0 +1,229 @@
+// it also tracks and updates epoch attesting balances.
+func ProcessEpochParticipation(
+	ctx context.Context,
+	state iface.BeaconState,
+	bal *precompute.Balance,
+	vals []*precompute.Validator,
+) ([]*precompute.Validator, *precompute.Balance, error) {
+	ctx, span := trace.StartSpan(ctx, "altair.ProcessEpochParticipation")
+	defer span.End()
+
+	cp, err := state.CurrentEpochParticipation()
+	if err != nil {
+		return nil, nil, err
+	}
+	for i, b := range cp {
+		if HasValidatorFlag(b, params.BeaconConfig().TimelyTargetFlagIndex) {
+			vals[i].IsCurrentEpochTargetAttester = true
+		}
+	}
+	pp, err := state.PreviousEpochParticipation()
+	if err != nil {
+		return nil, nil, err
+	}
+	for i, b := range pp {
+		if HasValidatorFlag(b, params.BeaconConfig().TimelySourceFlagIndex) {
+			vals[i].IsPrevEpochAttester = true
+		}
+		if HasValidatorFlag(b, params.BeaconConfig().TimelyTargetFlagIndex) {
+			vals[i].IsPrevEpochTargetAttester = true
+		}
+		if HasValidatorFlag(b, params.BeaconConfig().TimelyHeadFlagIndex) {
+			vals[i].IsPrevEpochHeadAttester = true
+		}
+	}
+	bal = precompute.UpdateBalance(vals, bal)
+	return vals, bal, nil
+}
+
+// ProcessRewardsAndPenaltiesPrecompute processes the rewards and penalties of individual validator.
+// This is an optimized version by passing in precomputed validator attesting records and and total epoch balances.
+func ProcessRewardsAndPenaltiesPrecompute(
+	state iface.BeaconStateAltair,
+	bal *precompute.Balance,
+	vals []*precompute.Validator,
+) (iface.BeaconStateAltair, error) {
+	// Don't process rewards and penalties in genesis epoch.
+	if helpers.CurrentEpoch(state) == 0 {
+		return state, nil
+	}
+
+	numOfVals := state.NumValidators()
+	// Guard against an out-of-bounds using validator balance precompute.
+	if len(vals) != numOfVals || len(vals) != state.BalancesLength() {
+		return state, errors.New("registries not the same length as state registries")
+	}
+
+	attsRewards, attsPenalties, err := AttestationsDelta(state, bal, vals)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not get attestation delta")
+	}
+
+	balances := state.Balances()
+	for i := 0; i < numOfVals; i++ {
+		vals[i].BeforeEpochTransitionBalance = balances[i]
+
+		// Compute the post balance of the validator after accounting for the
+		// attester and proposer rewards and penalties.
+		balances[i] = helpers.IncreaseBalanceWithVal(balances[i], attsRewards[i])
+		balances[i] = helpers.DecreaseBalanceWithVal(balances[i], attsPenalties[i])
+
+		vals[i].AfterEpochTransitionBalance = balances[i]
+	}
+
+	if err := state.SetBalances(balances); err != nil {
+		return nil, errors.Wrap(err, "could not set validator balances")
+	}
+
+	return state, nil
+}
+
+// AttestationsDelta computes and returns the rewards and penalties differences for individual validators based on the
+// voting records.
+func AttestationsDelta(state iface.BeaconStateAltair, bal *precompute.Balance, vals []*precompute.Validator) ([]uint64, []uint64, error) {
```

Comment:
Should this function be exported?

Reply:
For RPC end points. As good practice, I will not export them for now